File: 53. Maximum Subarray.java
/*
Problem Statement: https://leetcode.com/problems/maximum-subarray/
*/

class Solution {
    public int maxSubArray(int[] nums) {
        int maxSum = nums[0];
        int sum = 0;
        for (int num: nums) {
            sum += num;
            maxSum = Math.max(maxSum, sum);
            sum = Math.max(sum, 0);
        }
        return maxSum;
    }
}

File: Problems/53. Maximum Subarray.py
"""
Problem Statement: https://leetcode.com/problems/maximum-subarray/
"""


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        current_max_sum = nums[0]
        max_sum = nums[0]

        for i in range(1, len(nums)):
            current_max_sum = max(current_max_sum + nums[i], nums[i])
            max_sum = max(max_sum, current_max_sum)

        return max_sum
